modules/apps/headless/headless-builder/headless-builder-impl/src/main/java/com/liferay/headless/builder/internal/model/listener/HeadlessBuilderObjectEntryModelListener.java

```diff
@@ -0,0 +1,154 @@
+import com.liferay.object.model.ObjectDefinition;
+import com.liferay.object.model.ObjectEntry;
+import com.liferay.object.service.ObjectDefinitionLocalService;
+import com.liferay.object.service.ObjectEntryLocalService;
+import com.liferay.portal.kernel.exception.ModelListenerException;
+import com.liferay.portal.kernel.model.BaseModelListener;
+import com.liferay.portal.kernel.model.ModelListener;
+import com.liferay.portal.kernel.transaction.TransactionCommitCallbackUtil;
+import com.liferay.portal.kernel.util.StringUtil;
+
+import java.io.Serializable;
+
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
+
+import org.osgi.service.component.annotations.Component;
+import org.osgi.service.component.annotations.Reference;
+
+/**
+ * @author Sergio Jiménez del Coso
+ */
+@Component(service = ModelListener.class)
+public class HeadlessBuilderObjectEntryModelListener
+	extends BaseModelListener<ObjectEntry> {
+
+	@Override
+	public void onAfterCreate(ObjectEntry objectEntry)
+		throws ModelListenerException {
+
+		_schedulePublication(objectEntry);
+	}
+
+	@Override
+	public void onAfterUpdate(
+			ObjectEntry originalObjectEntry, ObjectEntry objectEntry)
+		throws ModelListenerException {
+
+		_schedulePublication(objectEntry);
+	}
+
+	private long _getAPIApplicationId(ObjectEntry objectEntry) {
+		if (_isObjectDefinition("L_API_APPLICATION", objectEntry)) {
+			return objectEntry.getObjectEntryId();
+		}
+		else if (_isObjectDefinition("L_API_SCHEMA", objectEntry)) {
+			Map<String, Serializable> values = objectEntry.getValues();
+
+			return (long)values.get(
+				"r_apiApplicationToAPISchemas_c_apiApplicationId");
+		}
+		else if (_isObjectDefinition("L_API_ENDPOINT", objectEntry)) {
+			Map<String, Serializable> values = objectEntry.getValues();
+
+			return (long)values.get(
+				"r_apiApplicationToAPIEndpoints_c_apiApplicationId");
+		}
+
+		throw new UnsupportedOperationException("Invalid object definition");
+	}
+
+	private boolean _isObjectDefinition(
+		String objectDefinitionExternalReferenceCode, ObjectEntry objectEntry) {
+
+		ObjectDefinition objectDefinition =
+			_objectDefinitionLocalService.fetchObjectDefinition(
+				objectEntry.getObjectDefinitionId());
+
+		return Objects.equals(
+			objectDefinition.getExternalReferenceCode(),
+			objectDefinitionExternalReferenceCode);
+	}
+
+	private void _schedulePublication(ObjectEntry objectEntry) {
+		ObjectDefinition objectDefinition =
+			_objectDefinitionLocalService.fetchObjectDefinition(
+				objectEntry.getObjectDefinitionId());
+
+		if (StringUtil.startsWith(
+				objectDefinition.getExternalReferenceCode(), "L_API_")) {
+
+			long apiApplicationId = _getAPIApplicationId(objectEntry);
+
+			_pendingAPIApplications.add(apiApplicationId);
+
+			TransactionCommitCallbackUtil.registerCallback(
+				() -> {
+					if (_pendingAPIApplications.contains(apiApplicationId)) {
```

Comment:
We could invoke directly `remove` method as it returns true if the element was removed:

`if (_pendingAPIApplications.remove(apiApplicationId)) {`

Reply:
Make sense @ccorreagg!!

Reply:
I like that.

Reply:
@sergiojimcos do it :)

Reply:
Done it!!

Reply:
https://github.com/liferay-headless/liferay-portal/pull/1321/commits/78cfb2a2138d38eb41e317cdf53caaeb2f444c5c